platform/ios/src/MGLFaux3DUserLocationAnnotationView.m

```diff
@@ -225,70 +215,67 @@ - (void)drawDot
-            _headingIndicatorLayer.contentsScale = [UIScreen mainScreen].scale;
-            _headingIndicatorLayer.opacity = 0.4;
-            _headingIndicatorLayer.shouldRasterize = YES;
-            _headingIndicatorLayer.rasterizationScale = [UIScreen mainScreen].scale;
-            _headingIndicatorLayer.drawsAsynchronously = YES;
-
-            [self.layer insertSublayer:_headingIndicatorLayer below:_dotBorderLayer];
+            [_headingIndicatorLayer removeFromSuperlayer];
+            _headingIndicatorLayer = nil;
+            _oldHeadingAccuracy = -1;
         }
 
-        // heading indicator accuracy mask (fan-shaped)
-        //
-        if ( ! _headingIndicatorMaskLayer && self.userLocation.heading.headingAccuracy)
+        if ( ! _headingIndicatorLayer && headingAccuracy)
         {
-            _headingIndicatorMaskLayer = [CAShapeLayer layer];
-            _headingIndicatorMaskLayer.frame = _headingIndicatorLayer.bounds;
-            _headingIndicatorMaskLayer.path = [[self headingIndicatorClippingMask] CGPath];
-
-            // apply the mask to the halo-radius-sized gradient layer
-            _headingIndicatorLayer.mask = _headingIndicatorMaskLayer;
-
-            _oldHeadingAccuracy = self.userLocation.heading.headingAccuracy;
-
+            if (headingTrackingModeEnabled)
+            {
+                _headingIndicatorLayer = [[MGLUserLocationHeadingBeamLayer alloc] initWithUserLocationAnnotationView:self];
+                [self.layer insertSublayer:_headingIndicatorLayer below:_dotBorderLayer];
+            }
+            else
+            {
+                _headingIndicatorLayer = [[MGLUserLocationHeadingArrowLayer alloc] initWithUserLocationAnnotationView:self];
+                [self.layer addSublayer:_headingIndicatorLayer];
+                _headingIndicatorLayer.zPosition = 1;
+            }
         }
-        else if (_oldHeadingAccuracy != self.userLocation.heading.headingAccuracy)
-        {
-            // recalculate the clipping mask based on updated accuracy
-            _headingIndicatorMaskLayer.path = [[self headingIndicatorClippingMask] CGPath];
 
-            _oldHeadingAccuracy = self.userLocation.heading.headingAccuracy;
+        if (_oldHeadingAccuracy != headingAccuracy)
+        {
+            [_headingIndicatorLayer updateHeadingAccuracy:headingAccuracy];
+            _oldHeadingAccuracy = headingAccuracy;
         }
 
         if (self.userLocation.heading.trueHeading >= 0)
         {
-            _headingIndicatorLayer.affineTransform = CGAffineTransformRotate(CGAffineTransformIdentity, -MGLRadiansFromDegrees(self.mapView.direction - self.userLocation.heading.trueHeading));
+            CGFloat rotation = -MGLRadiansFromDegrees(self.mapView.direction - self.userLocation.heading.trueHeading);
+
+            // Don't rotate if the change is imperceptible (arbitrarily chosen as: 0.01 radians/~0.6°).
+            if (fabs(rotation) > 0.01)
```

Comment:
Hoist this up top as a constant. Worth noting that this tolerance is much smaller than the other angular tolerance, `MGLToleranceForSnappingToNorth`, but I think that’s OK because the unrotation animation is meant to be perceptible.